lib/vanagon/component/dockerfile.rb

```diff
@@ -0,0 +1,188 @@
+    # structure, specifically in how they calculate the required commands that will
+    # need to run for a given step. That calculation logic should be factored out into
+    # a common module. Note that it is for precisely this reason that the individual
+    # "step" methods are essentially copy-pastes of the corresponding "rule" methods
+    # in Vanagon::Component::Rules.
+    class Dockerfile
+      include Vanagon::Utilities::ShellUtilities
+
+      # Create methods that generate the Dockerfile contents for a specific
+      # step when building the component.
+      #
+      # This method cuts some of the boilerplate command munging + Dockerfile
+      # generation for each step.
+      #
+      # The block should return an array of commands that will be run for a given
+      # step.
+      #
+      # @param step_name [String] The step's name
+      def self.step(step_name, &block)
+        define_method("#{step_name}_step") do
+          commands = instance_exec(&block)
+          commands.compact!
+          commands.reject!(&:empty?)
+          commands.flatten!
+          commands = andand_multiline(commands)
+
+          next "" if commands.empty?
+
+          <<-DOCKERFILE
+  # #{component.name}-#{step_name}
+  RUN #{commands}
+          DOCKERFILE
+        end
+      end
+
+      attr_accessor :component
+      attr_accessor :project
+      attr_accessor :platform
+
+      # @param component [Vanagon::Component] The component to generate the Dockerfile for.
+      # @param project [Vanagon::Project] The project associated with the component.
+      # @param platform [Vanagon::Platform] The platform where this component will be built.
+      def initialize(component, project, platform)
+        @component = component
+        @project = project
+        @platform = platform
+      end
+
+      # This override of andand_multiline is purely for aesthetic purposes, i.e. it makes
+      # the generated Dockerfile more readable.
+      def andand_multiline(*commands)
+        cmdjoin(commands, " && \\\n    ")
+      end
+
+      # Generate the component's Dockerfile
+      #
+      # @return [String] the contents of the component's Dockerfile.
+      def generate
+        steps = [
+          unpack_step,
+          patch_step,
+          configure_step,
+          build_step,
+          check_step,
+          install_step,
+        ]
+
+        if component.install_only
+          steps = [
+            install_step,
+          ]
+        end
+
+        steps.join("\n")
+      end
+
+      # Unpack the source for this component. The unpacking behavior depends on
+      # the source type of the component.
+      #
+      # @see [Vanagon::Component::Source]
+      step("unpack") do
+        [
+          andand_multiline(component.environment_variables, component.extract_with)
+        ]
+      end
+
+      # Apply any patches for this component.
+      step("patch") do
+        commands = []
```

Comment:
These are pretty much copy-pasted from their corresponding analogues in Component::Rules.

Reply:
If we want to go with this long term, it probably makes sense to factor some of the shared docker/make logic into a base class for output generation.